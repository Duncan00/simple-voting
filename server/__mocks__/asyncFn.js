'use strict';

const util = require('util');

const asyncFn = fname =>
	jest.fn(async (...args) => {
		const msg = `unexpected call to ${fname} with (${util.inspect(
			{args},
			10
		)}), did you miss any mockImplementationOnce?`;
		throw new Error(msg);
	});

module.exports = asyncFn;
