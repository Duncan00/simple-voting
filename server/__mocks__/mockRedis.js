'use strict';

const Redis = require('ioredis-mock');
const redis = new Redis();
// const asyncFn = require('./asyncFn');
//
// const redis = {
// 	mock_object_name: 'mock-redis-object',
// 	incr: asyncFn('redis#incr'),
// 	get: asyncFn('redis#get'),
// 	hmset: asyncFn('redis#hmset'),
// 	hgetall: asyncFn('redis#hgetall'),
// 	sismember: asyncFn('redis#sismember'),
// 	sadd: asyncFn('redis#sadd'),
// 	hincrby: asyncFn('redis#hincrby'),
// 	ping: asyncFn('redis#ping'),
// 	on: asyncFn('redis#on')
// };
const getRedis = jest.fn().mockResolvedValue(redis);
jest.doMock('../services/redis/getRedis', () => getRedis);

module.exports = redis;
