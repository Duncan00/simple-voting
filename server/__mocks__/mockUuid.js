'use strict';

// make uuid predictable in snapshots
jest.mock('uuid/v4', () => {
	let count = 0;
	return () => `uuid-${count++}`;
});
