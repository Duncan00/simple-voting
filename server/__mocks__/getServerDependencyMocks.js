'use strict';

/**
 *  Setup mocks for system boundaries (third party, data storage, etc) for easier testing
 */

const mockRedis = require('./mockRedis');

require('./mockUuid');

module.exports = {
	redis: mockRedis
};
