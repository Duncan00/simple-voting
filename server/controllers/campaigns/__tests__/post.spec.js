'use strict';

const supertest = require('supertest');
const initServer = () => require('../../../initServer')();
const getServerDependencyMocks = () =>
	require('../../../__mocks__/getServerDependencyMocks');

describe('POST /v1/campaigns', () => {
	let server;

	const campaign_record = {
		id: 'fa044192-db7e-4fd3-9334-c9344360286b',
		title: 'Who is the best NBA player in the history',
		start_date: '2019-01-01',
		end_date: '2019-01-31',
		'number_of_votes:Michael Jordan': '0',
		'number_of_votes:Tim Duncan': '0'
	};

	beforeEach(async () => {
		jest.resetModules();
		const {redis} = getServerDependencyMocks();
		redis.hmset.mockResolvedValueOnce(undefined);
		redis.hgetall.mockResolvedValueOnce(campaign_record);

		server = await initServer();
	});

	afterEach(() => {
		server.close();
	});

	describe('When all campaign names in request body are unique', () => {
		const request_body = {
			title: 'Who is the best NBA player in the history',
			start_date: '2019-01-01',
			end_date: '2019-01-31',
			candidates: [
				{
					name: 'Michael Jordan'
				},
				{
					name: 'Tim Duncan'
				}
			]
		};

		test('201 - Successfully created', async () => {
			const {body, status} = await supertest(server)
				.post(`/v1/campaigns`)
				.send(request_body);

			expect(body).toMatchSnapshot();
			expect(status).toEqual(201);
			const {redis} = getServerDependencyMocks();
			expect(redis.hmset.mock.calls).toMatchSnapshot('redis.hmset');
			expect(redis.hgetall.mock.calls).toMatchSnapshot('redis.hgetall');
		});
	});

	describe('When all campaign names in request body are not unique', () => {
		const request_body = {
			title: 'Who is the best NBA player in the history',
			start_date: '2019-01-01',
			end_date: '2019-01-31',
			candidates: [
				{
					name: 'Michael Jordan'
				},
				{
					name: 'Michael Jordan'
				}
			]
		};

		test('409 - Conflict', async () => {
			const {body, status} = await supertest(server)
				.post(`/v1/campaigns`)
				.send(request_body);

			expect(body).toMatchSnapshot();
			expect(status).toEqual(409);
		});
	});
});
