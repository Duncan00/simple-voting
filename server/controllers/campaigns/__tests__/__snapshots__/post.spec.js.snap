// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /v1/campaigns When all campaign names in request body are not unique 409 - Conflict 1`] = `
Object {
  "data": Object {},
  "meta": Object {
    "code": 40901,
    "errors": Array [],
    "message": "Candidate names are conflict.",
  },
}
`;

exports[`POST /v1/campaigns When all campaign names in request body are unique 201 - Successfully created 1`] = `
Object {
  "data": Object {
    "candidates": Array [
      Object {
        "name": "Michael Jordan",
        "number_of_votes": 0,
        "voted": false,
      },
      Object {
        "name": "Tim Duncan",
        "number_of_votes": 0,
        "voted": false,
      },
    ],
    "end_date": "2019-01-31",
    "id": "uuid-1",
    "start_date": "2019-01-01",
    "title": "Who is the best NBA player in the history",
  },
  "meta": Object {
    "code": 20100,
    "message": "Created",
  },
}
`;

exports[`POST /v1/campaigns When all campaign names in request body are unique 201 - Successfully created: hkid_candidate_set_ttls 1`] = `
Array [
  2735999,
  2735999,
]
`;

exports[`POST /v1/campaigns When all campaign names in request body are unique 201 - Successfully created: hkid_candidate_sets 1`] = `
Array [
  Array [
    "0",
  ],
  Array [
    "0",
  ],
]
`;

exports[`POST /v1/campaigns When all campaign names in request body are unique 201 - Successfully created: hkid_set 1`] = `
Array [
  "0",
]
`;

exports[`POST /v1/campaigns When all campaign names in request body are unique 201 - Successfully created: hkid_set_ttl 1`] = `2735999`;

exports[`POST /v1/campaigns When all campaign names in request body are unique 201 - Successfully created: inserted_campaign 1`] = `
Object {
  "end_date": "2019-01-31",
  "id": "uuid-1",
  "number_of_votes:Michael Jordan": "0",
  "number_of_votes:Tim Duncan": "0",
  "start_date": "2019-01-01",
  "title": "Who is the best NBA player in the history",
}
`;
